fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,201 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+public class LuceneSerializer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSerializer.class);
+    private static final int ENCODING_ENCRYPTED = 1;
+    private static final int ENCODING_COMPRESSED = 2;
+    private static final byte COMPRESSION_VERSION_FOR_HIGH_COMPRESSION = 0;
+
+    private static class EncodingState {
+        private boolean compressed;
+        private boolean encrypted;
+
+        private EncodingState() {
+            this.compressed = false;
+            this.encrypted = false;
+        }
+
+        private void setCompressed(boolean compressed) {
+            this.compressed = compressed;
+        }
+
+        private void setEncrypted(boolean encrypted) {
+            this.encrypted = encrypted;
+        }
+
+        private boolean isCompressed() {
+            return compressed;
+        }
+
+        private boolean isEncrypted() {
+            return encrypted;
+        }
+    }
+
+    @Nullable
+    public static byte[] encode(@Nullable byte[] data, boolean compress, boolean encrypt) {
+        if (data == null) {
+            return null;
+        }
+
+        final EncodingState state = new EncodingState();
+        final ByteBuffersDataOutput decodedDataOutput = new ByteBuffersDataOutput();
+        // Placeholder for the code byte at beginning, will be modified in output byte array if needed
+        decodedDataOutput.writeByte((byte) 0);
+
+        byte[] encoded = compressIfNeeded(compress, data, decodedDataOutput, state, 1);
+        encoded = encryptIfNeeded(encrypt, encoded, state, 1);
+
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        LOGGER.trace("Encoded data for compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+
+        byte[] decoded = decryptIfNeeded(encrypted, data, 1);
+        decoded = decompressIfNeeded(compressed, decoded, 1);
```

Comment:
This is a weird contract, as though `decryptIfNeeded` had to decrypt after offset bytes but still return them in the clear text.

Would it make sense to just remove encryption for now in the interest of getting compression done and revisit later, with whatever fallout that has to the API?

Reply:
yeah it is weird that the first byte was still kept after decryption. Removed the encryption/decryption methods.